### PR TITLE
feat: add auth config for sas9 in target

### PR DESCRIPTION
**Pull request by sabhas**
## Intent

The intention is to add authconfig for sas9 server type add encoding decoding utility.

## Coverage Report
Screenshot from 2021-07-02 14-41-48

## Checks

- [x] Code is formatted correctly (`npm run lint:fix`).
image
- [x] All unit tests are passing (`npm test`).
image
- [ ] All `sasjs-cli` unit tests are passing (`npm test`).
- [x] Reviewer is assigned.

### CODE REVIEW

File: src/utils/base64.spec.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import { encodeToBase64, decodeFromBase64 } from './base64'
+
+describe('base64', () => {
+  it('should return a base64 encoded string with encoded as prefix', () => {
```
Comment (allanbowe) [inline src/utils/base64.spec.ts:4]: Can we have a longer prefix - perhaps 100 characters?

File: src/types/config.ts [hunk 2]
```diff
@@ -50,6 +50,11 @@ export interface AuthConfig {
   secret: string
 }
 
+export interface AuthConfigForSas9 {
```
Comment (YuryShkoda) [inline src/types/config.ts:53]: pls rename it to `AuthConfigSAS9 `

File: src/types/target.spec.ts [hunk 3]
```diff
@@ -310,6 +310,29 @@ describe('Target', () => {
...
+      authConfigForSas9
+    })
+
+    expect(target).toBeTruthy()
+    expect(target instanceof Target).toEqual(true)
+    expect(target.name).toEqual('test')
+    expect(target.serverUrl).toEqual('')
+    expect(target.serverType).toEqual(ServerType.Sas9)
+    expect(target.appLoc).toEqual('/test')
+    expect(target.authConfigForSas9).toEqual(authConfigForSas9)
```
Comment (YuryShkoda) [inline src/types/target.spec.ts:332]: you only need this assertion

File: src/types/target.spec.ts [hunk 4]
```diff
@@ -310,6 +310,29 @@ describe('Target', () => {
...
+  })
+
+  //224,234,246,256,266,277,285
```
Comment (YuryShkoda) [inline src/types/target.spec.ts:335]: please remove unnecessary comments

File: src/types/target.spec.ts [hunk 5]
```diff
@@ -392,6 +415,40 @@ describe('Target', () => {
     expect(json.deployConfig).toEqual(undefined)
   })
 
+  it('should convert to JSON with authConfig is equal to this.authConfig', () => {
```
Comment (YuryShkoda) [inline src/types/target.spec.ts:418]: please clarify the purpose of this test case

File: src/types/targetValidators.ts [hunk 6]
```diff
@@ -139,6 +140,18 @@ export const validateAuthConfig = (authConfig: AuthConfig): AuthConfig => {
   return authConfig
 }
 
+export const validateAuthConfigForSas9 = (
```
Comment (YuryShkoda) [inline src/types/targetValidators.ts:143]: please rename it to `validateAuthConfigSAS9 `

Comment (YuryShkoda) [reply src/types/targetValidators.ts:143]: and you should also validate that `userName` and `password` are not empty strings

File: src/utils/base64.spec.ts (hunk 1 above)
Comment (YuryShkoda) [inline src/utils/base64.spec.ts:4]: I would create a variable for a test string

File: src/types/targetValidators.spec.ts [hunk 7]
```diff
@@ -657,25 +659,48 @@ describe('validateRepositoryName', () => {
...
-      validateAuthConfig({
+  describe('For SASVIYA', () => {
+    it('should throw an error when authConfig is null', () => {
+      expect(() =>
+        validateAuthConfig(null as unknown as AuthConfig)
+      ).toThrowError('Invalid auth config: JSON cannot be null or undefined.')
+    })
+
+    it('should return the auth config when valid', () => {
+      expect(
```
Comment (YuryShkoda) [inline src/types/targetValidators.spec.ts:670]: you should create a variable for the config

File: src/types/targetValidators.spec.ts [hunk 8]
```diff
@@ -657,25 +659,48 @@ describe('validateRepositoryName', () => {
...
         secret: '53CR3T'
       })
-    ).toEqual({
-      access_token: 'T35T',
-      refresh_token: 'R3FR35H',
-      client: 'CL13NT',
-      secret: '53CR3T'
+    })
+  })
+  describe('For SAS(', () => {
```
Comment (YuryShkoda) [inline src/types/targetValidators.spec.ts:685]: delete `(`

File: src/types/targetValidators.spec.ts [hunk 9]
```diff
@@ -657,25 +659,48 @@ describe('validateRepositoryName', () => {
...
+    it('should throw an error when authConfigForSas9 is null', () => {
+      expect(() =>
+        validateAuthConfigForSas9(null as unknown as AuthConfigForSas9)
+      ).toThrowError(
+        'Invalid auth config for sas9: JSON cannot be null or undefined.'
+      )
+    })
+
+    it('should return the auth config for sas9 when valid', () => {
```
Comment (YuryShkoda) [inline src/types/targetValidators.spec.ts:694]: you should create a variable for the config

**Review by YuryShkoda — CHANGES_REQUESTED** [review]

**Comment by sabhas** [comment]
Screenshot from 2021-07-01 21-52-10

**Review by VladislavParhomchik — APPROVED** [review]

**Review by allanbowe — APPROVED** [review]

**Pull request merged**